presto/dataset.py

```diff
@@ -209,11 +211,33 @@ def __init__(
             dataframe["end_date"] = pd.to_datetime(dataframe.end_date)
             dataframe = dataframe[(~dataframe.end_date.dt.year.isin(years_to_remove))]
         self.target_function = target_function if target_function is not None else self.target_crop
+        self._class_weights: Optional[np.ndarray] = None
+
         super().__init__(dataframe)
+        if balance:
+            neg_indices, pos_indices = [], []
+            for loc_idx, (_, row) in enumerate(self.df.iterrows()):
+                target = self.target_function(row.to_dict())
+                if target == 0:
+                    neg_indices.append(loc_idx)
+                else:
+                    pos_indices.append(loc_idx)
+            if len(pos_indices) > len(neg_indices):
+                self.indices = pos_indices + (len(pos_indices) // len(neg_indices)) * neg_indices
+            elif len(neg_indices) > len(pos_indices):
+                self.indices = neg_indices + (len(neg_indices) // len(pos_indices)) * pos_indices
```

Comment:
so this is implicitly leading to the batch balancing? Is this afterwards still shuffled?

Reply:
Yes - this is still shuffled so a batch is not necessarily balanced (but the dataset as a whole will be mostly balanced).

The shuffling is done afterwards, by the dataloader